README.md

```diff
@@ -1,115 +1,82 @@
-<!-- START_HASH_TRAITS -->
-1. `WriteHash()`: writes the object to a binary format using `StableHashTraits.write(io, x)`
-    and takes a hash of that. `StableHashTraits.write(io, x)`
-    falls back to `Base.write(io, x)` if no specialized methods are defined for x.
-2. `IterateHash()`: assumes the object is iterable and finds a hash of all elements
-3. `StructHash([pair = (fieldnames ∘ typeof) => getfield], [order])`: hash the structure of
-    the object as defined by a sequence of pairs. How precisely this occurs is determined by
-    the two arguments:
-      - `pair` Defines how fields are extracted; the default is
-        `fieldnames ∘ typeof => getfield`
-        but this could be changed to e.g. `propertynames => getproperty` or
-        `Tables.columnnames => Tables.getcolumn`. The first element of the pair is a
-        function used to compute a list of keys and the second element is a two argument
-        function used to extract the keys from the object.
-      - `order` can be `:ByOrder` (the default)—which sorts by the order returned by
-        `pair[1]`—or `:ByName`—which sorts by lexigraphical order.
-4. `FnHash(fn, [method])`: hash the result of applying `fn` to the given object. Optionally,
-   use `method` to hash the result of `fn`, otherwise calls `hash_method` on the result to
-   determine how to hash it. There are two built-in functions commonly used with
-   `FnHash`.
-    - `stable_typename_id`: Get the qualified name of an object's type, e.g. `Base.String` and return 64 bit hash of this string
-    - `stable_type_id`: Get the qualified name and type parameters of a type, e.g.
-       `Base.Vector{Int}`, and return a 64 bit hash of this string.
-5. `@ConstantHash(x)`: at compile time, hash the literal (constant) string or number using
-    `sha256` and include the first 64 bits as a constant number that is recursively hashed
-    using the `WriteHash` method.
-6. `Tuple`: apply multiple methods to hash the object, and then recursively hash their
-    results. For example: `(@ConstantHash("header"), StructHash())` would compute a hash for
-    both the string `"header"` and the fields of the object, and then recursively hash
-    these two hashes.
-
-Your hash will be stable if the output for the given method remains the same: e.g. if
-`write` is the same for an object that uses `WriteHash`, its hash will be the same; if the
-fields are the same for `StructHash`, the hash will be the same; etc...
-
-Missing from the above list is one final, advanced, trait: `HashAndContext` which can be used to change the context within the scope of a given object. You can learn more about contexts below.
-
-<!-- END_HASH_TRAITS -->
+- `StructType.ArrayType` — the `eltype` is hashed and elements are hashed using `iterate`
+
+- `StructType.DictType` — the `eltype` of the keys and values are hashed by iterating over `StructTypes.keyvaluepairs`
```

Comment:
I wonder, do we need to care about the element type? I haven't looked through everything yet but it looks like it adds complexity in the `union_split` methods. To me it seems OK if `Any[1,2,3]` and `Int[1,2,3]` hash the same, as long as we document it.

Reply:
This is something I spent a while going back and forth about when I was working on this.

Note that as implemented, e.g. `Int[1,2,3]` and `UInt[1,2,3]` would hash to the same value, since the name hashed for the eltype is `NumberType` and the byte sequence is identical. However `Int32[1,2,3]` and `Char[Char(1), Char(2), Char(3)]` is different, since the the name hashed is `NumberType` for the integers and `StringType` for the characters (and even though the byte sequence of the array itself is identical). The actual hash of a type is determined by that first bullet (which hashes according to `StructType(T)`).

So long as we want to differentiate between some types of elements (e.g. array of some complex struct vs array of numbers) hashing the eltype/fieldtypes makes it easier to write an efficient implementation.
 
This is because I use the assumption that the eltype and fieldtypes of a container type is hashed to type-hoist: e.g. if I have a concrete eltype I only need hash the content of the elements not the type of each element. Because I only need to check that the type is concrete (rather than checking whether the containing type has hashed something that include the current contained elements) it is easy to let such hoisting propagate over nested structures and arrays, avoiding the type hash for all the contained components so long as they are concrete. In my benchmarks the type-hoisting optimization leads to roughly an order of magnitude or more speedup.

Reply:
hm, should `Any[Point(0,0), Point(0,0)]` has the same as `Point[Point(0,0), Point(0,0)]`?

Reply:
ah I think I see what you are saying. If we required they hash the same, then the hash has to be obtained without hoisting (since we can't really hoist in the first case), so we get the same results either way. but if we allow the hash to depend on the eltype, then we can have different hashes for these two, which lets us hoist the type hash in the second case.

Reply:
I guess it would be possible to compute some sort of "defacto" eltype in the first case, which would be slow but would then let you "type-hoist" (to no optimization) effect, which would ensure that the two hashed the same. That feels like a lot of engineering to achieve what I'm not sure we *need* to be true.